ENH: tpx 137 (GROMACS 2025.0) support

tylerjereddy opened the pull request:
* Add support for GROMACS `2025.0` `.tpr` version 137, which mostly involves accounting for the new PyTorch neural network potential field, `F_ENNPOT`. Most of the other changes are similar to the approach I previously used for bumping `.tpr` version support in gh-4866.

* Some other `.tpr` additions are noted at https://github.com/MDAnalysis/mdanalysis/discussions/4888, but we don't appear to need to use those within the context of our current `.tpr` parsing tests. It may be sensible to occasionally generate `.tpr` test files that use those extra bells and whistles to see how we fare, though as a first pass here the full testsuite passes for us if we simply do the usual `gmx convert-tpr` at the latest stable GROMACS release tag `v2025.0` and propagate our current TPR testing to additionally use the converted files.


## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [ ] Issue raised/referenced?
 - [x] Tests updated/added?
 - [x] Documentation updated/added?
 - [x] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4919.org.readthedocs.build/en/4919/

<!-- readthedocs-preview mdanalysis end -->

tylerjereddy committed:
ENH: tpx 137 support

* Add support for GROMACS `2025.0` `.tpr` version 137, which mostly
involves accounting for the new PyTorch neural network potential
field, `F_ENNPOT`. Most of the other changes are similar to the
approach I previously used for bumping `.tpr` version support
in gh-4866.

* Some other `.tpr` additions are noted at https://github.com/MDAnalysis/mdanalysis/discussions/4888,
but we don't appear to need to use those within the context of our
current `.tpr` parsing tests. It may be sensible to occasionally
generate `.tpr` test files that use those extra bells and whistles to
see how we fare, though as a first pass here the full testsuite passes
for us if we simply do the usual `gmx convert-tpr` at the latest stable
GROMACS release tag `v2025.0` and propagate our current TPR testing
to additionally use the converted files.

tylerjereddy commented on package/MDAnalysis/topology/tpr/utils.py:
```diff
@@ -702,6 +702,10 @@ def do_iparams(data, functypes, fver):
+        elif i in [setting.F_ENNPOT]:
+            # TODO: handle the new neural network potentials
+            # supported from GROMACS 2025.0
+            pass
```
For this and other new `v2025.0` bells and whistles, cc @al42and for possible generation of new test files that have the new bells and whistles.

I don't now if we really need a new test file with i.e., the new PyTorch potential handling + reference coordinate scaling with multiple centers of mass + init-histogram-counts to enable continuation of expanded ensemble runs.

Probably would be good to have eventually, if only to make sure we don't need to skip over any other bits of information when "seeking" through the file.

IAlibay replied:
Do we think there are attributes we will be missing if we're not properly parsing the NN potential things? i.e. what I'm asking is if we need a warning here.

al42and replied:
> For this and other new v2025.0 bells and whistles, cc @al42and for possible generation of new test files that have the new bells and whistles.

I can make a TPR with a bunch of things turned on, but there are issues if we want to stick with existing test sets:
- 2lyz: they only have TPR, which is neither editable nor "decompilable" into parts; can't change settings there (and looks like the original receipt is lost in time, since the new are just converted from the old ones, per 336ff1d5238b1bfaa1d9a38fd142db54e35d5104).
- bonded/dummy: currently, for NN, we need something with "physical" element names.

If that's ok, I can create a new TPR roughly based on GROMACS's own tests.
Reconstructing 2lyz and making a test based on it is doable, but I'd rather avoid that as long as you're ok with adding a new test system.

IAlibay reviewed (approved):
I'm not very versed in the new 2025 changes, but this looks reasonable to me, thanks!

tylerjereddy commented:
> Reconstructing 2lyz and making a test based on it is doable, but I'd rather avoid that as long as you're ok with adding a new test system.

@al42and I think making a new `.tpr` unrelated to our current testing `.tpr`s would still be useful if it has the new bells and whistles turned on, especially if you could keep the file size suitable for a testsuite (i.e., similar file size range to what we currently use). Presumably we could at least use `gmx dump` to get some ground truths for sanity checking our parser.

al42and committed:
TST: Add GROMACS 2025.0 TPR with NNpot

* Generated input file with random set of features enabled
Using inputs from gromacs/src/testutils/simulationdatabase/
`gmx grompp -f tt6710474.mdp -c  ala.gro -n ala.ndx -p ala.top  -o ala_nnpot_gmx_2025_0.tpr`

* Add Andrey to `CHANGELOG`

al42and commented:
Hi! 

Here's an example file + test. The NN potential is stored as a reference, so there's not really much to test except that it does not screw up anything else. Also enabled a few other random features in the MDP. The file size is 11KiB (that's just ALA dipeptide in vacuum)

https://github.com/MDAnalysis/mdanalysis/commit/e6e8bb56114d656bcdf7cd3a2f5e4b96db4814fd

orbeckst reviewed (approved):
Having the capability to read the basic features for the new format (as demonstrated in the PR on the converted TPR files) is "good enough" for getting this into the 2.9.0 release #4939 . Given the time constraints I'd suggest to follow up with further work in additional PRs.

If we're worried that other things really won't work well then we could add a note to docs/release notes. @tylerjereddy I'll leave this to your judgement.

tylerjereddy commented:
I'm at the airport in Austin, but just pulled in Andrey's commit and the testsuite seems to be happy locally even with the fancy new TPR, so I went ahead and pushed that up alongside Andrey's addition to the CHANGELOG.

So, no known reservations on my end, may even be able to remove the residual `TODO`, though that doesn't hurt

IAlibay commented:
Amazing, thanks @al42and and @tylerjereddy !

IAlibay commented on testsuite/MDAnalysisTests/topology/test_tprparser.py:
```diff
@@ -205,6 +211,18 @@ class TestTPR46x(TPRAttrs):
     def filename(self, request):
         return request.param
 
+class TestTPRNnpot(TPRAttrs):
```
```suggestion

class TestTPRNnpot(TPRAttrs):
```

IAlibay commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -612,6 +621,10 @@
     _data_ref / "tprs/virtual_sites/extra-interactions-4.0.7.tpr"
 ).as_posix()
+# ALA dipeptide with neural network potential and a few other options
+TPR_NNPOT_2025_0 = (
+    _data_ref / "tprs/ala_nnpot_gmx_2025_0.tpr"
+).as_posix()
```
```suggestion
TPR_NNPOT_2025_0 = (_data_ref / "tprs/ala_nnpot_gmx_2025_0.tpr").as_posix()
```

IAlibay reviewed (commented):
black linting (couldn't push from the CLI for some reason)

IAlibay committed:
Apply suggestions from code review

IAlibay commented:
CI was passing earlier and now that linters are I'm going to go ahead with the merge.
